### PR TITLE
feat: Snyk token submission modal

### PR DESCRIPTION
**Pull request by mathur07**
**SS#1**
image

**SS#2**
image

**SS#3**
image

**SS#4**
image

**SS#5**
image

**SS#6**
image

### CODE REVIEW

File: src/app/stack/stack-details/stack-details.component.ts [hunk 1]
```diff
@@ -227,7 +234,100 @@ export class StackDetailsComponent implements OnChanges {
...
+    tokenAlertsMessage: string;
+
+
+    observableToken: Observable<any>;
+
+    constructor(private stackAnalysisService: StackAnalysesService, private modalService: BsModalService) { }
+
+    async submitToken() {
+        if (!this.tokenErrorStatus) {
+            await this.stackAnalysisService.putToken(this.getBaseUrl(this.stack), '4c2c13be-2f0b-44a0-9e86-18abb3c60a6b', this.token, this.gatewayConfig)
```
Comment (arajkumar) [inline src/app/stack/stack-details/stack-details.component.ts:258]: Why `4c2c13be-2f0b-44a0-9e86-18abb3c60a6b` is hardcoded?

File: src/app/stack/stack-details/stack-details.component.ts [hunk 2]
```diff
@@ -227,7 +234,100 @@ export class StackDetailsComponent implements OnChanges {
...
+                this.init()
+            }, 5000);
+        }
+
+    }
+
+    async setTokenStatus() {
+
+        // 3620266d-ae88-42ce-bbb0-699f2dd67c49
+        await this.stackAnalysisService.getTokenStatus(this.getBaseUrl(this.stack), '4c2c13be-2f0b-44a0-9e86-18abb3c60a6b', this.gatewayConfig)
```
Comment (arajkumar) [inline src/app/stack/stack-details/stack-details.component.ts:277]: Why token is hardcoded?

File: src/app/stack/stack-details/stack-details.component.ts [hunk 3]
```diff
@@ -227,7 +234,100 @@ export class StackDetailsComponent implements OnChanges {
...
+            length: 0
+        };
+        this.tokenerror.length = this.token.length;
+        if (this.token.length === 0 || this.token == undefined) {
+            this.tokenerror.status = true;
+            this.tokenerror.type = '';
+        } else if (this.token.length !== 36) {
+            this.tokenerror.status = true;
+            this.tokenerror.type = "Token length should be of 36 character";
+        } else if (!/^[a-zA-Z0-9-]*$/.test(this.token)) {
```
Comment (arajkumar) [inline src/app/stack/stack-details/stack-details.component.ts:320]: UUID is a hex string. Could you follow https://stackoverflow.com/a/13653180/1942688 for the proper regex?

File: src/app/stack/stack-analyses.service.ts [hunk 4]
```diff
@@ -63,6 +63,53 @@ export class StackAnalysesService {
...
+          }
+        })
+          .map(this.extractTokenData)
+          .toPromise()
+      }
+    }
+    return null;
+  }
+
+  putToken(url: string, uuid: string, token: string, params?: any) {
```
Comment (arajkumar) [inline src/app/stack/stack-analyses.service.ts:88]: Can this function be verbose? something like `linkSynkTokenWithUserID`?

File: src/app/stack/stack-analyses.service.ts [hunk 5]
```diff
@@ -63,6 +63,53 @@ export class StackAnalysesService {
     return null;
   }
 
+  getTokenStatus(url: string, uuid: string, params?: any) {
+    let getURL = url.concat('user/', uuid);
+    if (params) {
+      if (params['access_token']) {
```
Comment (arajkumar) [inline src/app/stack/stack-analyses.service.ts:69]: Can you use optional chaining?
`params?.access_token`

Comment (arajkumar) [reply src/app/stack/stack-analyses.service.ts:69]: Use this pattern where ever applicable in this PR.

**Review by arajkumar — CHANGES_REQUESTED** [review]
Where is the fallback logic for v2?

File: src/app/stack/stack-analyses.service.ts [hunk 6]
```diff
@@ -87,6 +133,13 @@ export class StackAnalysesService {
     return body as StackReportModel;
   }
 
+  private extractTokenData(res: Response) {
+    let body = res.json() || {};
+    body['statusCode'] = res.status;
+    body['statusText'] = res.statusText;
+    return body as TokenDetailModel;
+  }
```
Comment (arajkumar) [inline src/app/stack/stack-analyses.service.ts:141]: Make this as lambda as it doesn't seem to be used elsewhere.

**Review by arajkumar — CHANGES_REQUESTED** [review]

File: src/app/stack/stack-details/stack-details.component.html [hunk 7]
```diff
@@ -1,40 +1,121 @@
...
+                                &nbsp;
+                                <button type="button" class="pf-c-button pf-m-small pf-m-link"
+                                    (click)="modalRef.hide()">Close</button>
+                            </footer>
+                        </div>
+                    </ng-template>
+                </div>
+                <div class="col-md-1">
+                    <button type="button" class="btn btn-primary btn-sm">Share</button>
+                </div>
```
[the remark this answers is not in the document]
Comment (mathur07) [reply src/app/stack/stack-details/stack-details.component.html]: Done

File: src/app/stack/stack-details/stack-details.component.ts (hunk 1 above)
Comment (mathur07) [reply src/app/stack/stack-details/stack-details.component.ts]: removed

File: src/app/stack/stack-details/stack-details.component.ts (hunk 2 above)
Comment (mathur07) [reply src/app/stack/stack-details/stack-details.component.ts]: removed

File: src/app/stack/stack-details/stack-details.component.ts (hunk 3 above)
Comment (mathur07) [reply src/app/stack/stack-details/stack-details.component.ts]: updated it.

File: src/app/app.component.ts [hunk 8]
```diff
@@ -62,13 +63,16 @@ export class AppComponent implements OnInit {
                     this.gateway['config']['api_url'] = apiHost;
                 }
 
-                // Separate v2 users 
-                if (this.gateway['config']['ver'] === 'v2') {
+                // Separate v1, v2, v3 users 
+                if (this.gateway['config']['ver'] === 'v3') {
+                    this.version = 'v2';
+                    this.uuid = this.gateway['config']['uuid'];
+                } else if (this.gateway['config']['ver'] === 'v2') {
```
Comment (mathur07) [inline src/app/app.component.ts:70]: @arajkumar this is how we are managing fallback to v2. With this, we set UUID and then use this UUID to show the Snyk Token button

File: src/app/stack/stack-analyses.service.ts (hunk 4 above)
Comment (mathur07) [reply src/app/stack/stack-analyses.service.ts]: Updated it

File: src/app/stack/stack-analyses.service.ts (hunk 6 above)
Comment (mathur07) [reply src/app/stack/stack-analyses.service.ts]: updated it

File: src/app/stack/stack-details/stack-details.component.ts [hunk 9]
```diff
@@ -227,7 +234,104 @@ export class StackDetailsComponent implements OnChanges {
...
+        } else if (!/^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/.test(this.token)) {
+            this.tokenerror.status = true;
+            this.tokenerror.type = "Token can only have [a-zA-Z0-9-]";
+        }
+
+        if (this.tokenerror.status) {
+            this.tokenErrorStatus = true;
+        } else {
+            this.tokenErrorStatus = false;
+        }
```
Comment (arajkumar) [inline src/app/stack/stack-details/stack-details.component.ts:333]: Can it be rearranged like below?
```suggestion
        this.tokenErrorStatus = this.tokenerror.status===true;
```

File: src/app/stack/stack-details/stack-details.component.ts [hunk 10]
```diff
@@ -385,42 +489,29 @@ export class StackDetailsComponent implements OnChanges {
...
-                                            'You don\'t seem to have sufficient privileges to access this';
-                                    }
-                                    title = 'Report failed'; // Check if just this message is enough.
-                                    this.handleError({
-                                        message: error.statusText,
-                                        status: error.status,
-                                        title: title
-                                    });
-                                });
+                this.setTokenStatus()
```
Comment (arajkumar) [inline src/app/stack/stack-details/stack-details.component.ts:492]: Remove the counter declaration as well

File: src/app/stack/stack-details/stack-details.component.ts [hunk 11]
```diff
@@ -227,7 +234,104 @@ export class StackDetailsComponent implements OnChanges {
...
+                    console.log(res.status);
+                })
+                .catch(err => {
+                    console.log(err);
+                });
+
+            console.log('refresh in 5 sec');
+            setTimeout(() => {
+                this.init()
+            }, 5000);
```
Comment (arajkumar) [inline src/app/stack/stack-details/stack-details.component.ts:269]: What is the logic behind this 5sec?

File: src/index.html [hunk 12]
```diff
@@ -4,6 +4,7 @@
     <title>Fabric8 Stack Analysis</title>
     <meta charset="UTF-8">
     <meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport">
+    <link rel="stylesheet" href="https://unpkg.com/@patternfly/patternfly/patternfly.css" crossorigin="anonymous">
```
Comment (arajkumar) [inline src/index.html:7]: You can add a comment saying that we just use css declaration from patternfly

**Review by arajkumar — CHANGES_REQUESTED** [review]

File: src/app/stack/stack-details/stack-details.component.ts (hunk 11 above)
Comment (mathur07) [reply src/app/stack/stack-details/stack-details.component.ts:269]: Its just a wait time of 5 sec. Just to achieve consistency on the getCall inside init().

File: src/app/stack/stack-details/stack-details.component.ts (hunk 10 above)
Comment (mathur07) [reply src/app/stack/stack-details/stack-details.component.ts:508]: Done

File: src/app/stack/stack-details/stack-details.component.ts (hunk 9 above)
Comment (mathur07) [reply src/app/stack/stack-details/stack-details.component.ts]: Yes. Done

File: src/index.html (hunk 12 above)
Comment (mathur07) [reply src/index.html:8]: Done

**Review by arajkumar — APPROVED** [review]
lgtm

**Pull request merged**